### PR TITLE
[lodash] Extend typescript support of lodash.isPlainObject function

### PR DESCRIPTION
**Pull request by rosolko**
Initially this changes inspired by impelentation of `_.isSet` function.

When I'm using typescript each time after I check raw object with `_.isPlainObject` function inside operator I should cast raw object to plain by `Record<string, unknown>` to properly work with. 

Sample, this case I can't access `foo` property because of non fully check after which typesctipr doesn't really know which type the object is:

```typescript
const rawObject = {foo: 'bar'};
if (_.isPlainObject(rawObject)) {
	rawObject.f̼o̼o̼ = 'baz';
}
```

To make it work I should made a cast and this case accessing `foo` property is possible:

```typescript
const rawObject = {foo: 'bar'};
if (_.isPlainObject(rawObject)) {
	const typedObject = rawObject as Record<string, unknown>;
	typedObject.foo = 'baz';
}
```

Proposed changes change allow to cast raw object ot plain on a check and easily access `foo` property with such construction:

```typescript
const rawObject = {foo: 'bar'};
if (_.isPlainObject(rawObject)) {
	rawObject.foo = 'baz';
}
```

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://www.typescriptlang.org/docs/handbook/utility-types.html#recordkeys-type
- [x] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.

### CODE REVIEW

File: types/lodash/common/lang.d.ts [hunk 1]
```diff
@@ -1114,7 +1114,7 @@ declare module "../index" {
          * @param value The value to check.
          * @return Returns true if value is a plain object, else false.
          */
-        isPlainObject(value?: any): boolean;
+        isPlainObject(value?: any): value is Record<string, unknown>;
```
Comment (aj-r) [inline types/lodash/common/lang.d.ts:1117]: This is not accurate in all cases and could cause issues in certain scenarios (especially in the else block).

**Review by aj-r — COMMENTED** [review]
I don't fully understand the motivation for this. Your example compiles for me even without this change. I assume that the value actually has `unknown` type?

In the case of `unknown`, I don't think `isPlainObject` or `isObject` should magically grant access to all properties that may or may not exist. Those properties may also exist on other types (even strings and numbers!) so there is nothing special about being an object. The point of `unknown` is that you don't know what properties exist, so you need to cast if you want access to them. So I think casting is the desired behavior here.

There is a similar discussion for `isObject` in #59138.

File: types/lodash/common/lang.d.ts (hunk 1 above)
Comment (rosolko) [reply types/lodash/common/lang.d.ts:1117]: As far as I know _plain_ object in typecript just the same there is a `Record<string, unknown>` so why can't we check like that?

Comment (rosolko) [reply types/lodash/common/lang.d.ts:1117]: I think here you should check for a specific class type with another is or just throw an error about unknown type (for example)

**Review by RyanCavanaugh — CHANGES_REQUESTED** [review]
@aj-r's comment is correct; a type guard must return `true` for *every* value in its tested domain but this function does not do that.

**Pull request closed without merging**